### DISCUSSION TITLE
How many files per repository folder? Is 1000 the max?  I got an error on my repo saying 'Sorry, we had to truncate this directory to 1,000 files. 170 entries were omitted from the list' at the home page?

### DISCUSSION THREAD

**Discussion by jlpabon** [community/community#136892] [New to GitHub] [answered]
Dear all,

How many files per repository folder is the maximum? Is 1000 the max?  I got an error on my repo saying 'Sorry, we had to truncate this directory to 1,000 files. 170 entries were omitted from the list' at the home page?  Are the files there but not listed, or just not there?

The whole directory is 420 MB, which I believe is under the suggested guidelines. ( I found some threads about the file size).

The repository is private.

Thanks for helping me learn github.

Best wishes,
-j

**Comment by zshaian** [comment, marked as answer]
Yeah, GitHub does indeed have a limit on the number of files it can display in a single folder. The message you're seeing, "Sorry, we had to truncate this directory to 1,000 files. 170 entries were omitted from the list," indicates that the folder contains more than 1,000 files. GitHub will only display the first 1,000 files, and any additional files won't be listed in the UI, but they're still there in the repository.

If you need to access all the files from the GitHub web interface frequently, consider reorganizing your repository structure into subfolders to keep the number of files per folder below the 1,000 file limit.

Hope this helps!

**Comment by jlpabon** [comment]
Thanks, excellent information. It is great to know the files are still there.  In the near future, I will see about automating some data into subfolders...

**Comment by smoyer64** [comment]
Painfully, this also seems to apply to the API as there are not pagination links provided - e.g. https://api.github.com/repos/ethereum-lists/chains/contents/_data/chains.  I also tried setting `per_page` to a very high number with no effect.

**Reply by kyanny** [reply]
@smoyer64
[Get repository content](https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#get-repository-content) API has an upper limit of 1,000 files for a directory. If you need to retrieve more files, use the [Git Trees API](https://docs.github.com/rest/git/trees#get-a-tree).

1. [Get a reference](https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#get-a-reference)

    ```
    ❯ gh api repos/ethereum-lists/chains/git/ref/heads/master -q '.object.sha'
    21ccd03e9dd0fc16a14a6902bac2814b9b7159bb
    ```

2. [Get a commit object](https://docs.github.com/en/rest/git/commits?apiVersion=2022-11-28#get-a-commit-object)

    ```
    ❯ gh api repos/ethereum-lists/chains/git/commits/21ccd03e9dd0fc16a14a6902bac2814b9b7159bb -q '.tree.sha'
    fa19661238a17a36ee88e7cd9afebfa70becb5dc
    ```

3. [Get a tree](https://docs.github.com/en/rest/git/trees?apiVersion=2022-11-28#get-a-tree)

    ```
    ❯ gh api repos/ethereum-lists/chains/git/trees/fa19661238a17a36ee88e7cd9afebfa70becb5dc -q '.tree[] | select(.path == "_data") | .sha'
    6b2aff99537a3a752c7618eabab92a6cdefb389c
    
    ❯ gh api repos/ethereum-lists/chains/git/trees/6b2aff99537a3a752c7618eabab92a6cdefb389c -q '.tree[] | select(.path == "chains") | .sha'
    4149a134bc500908ef28f43d32601b119473454b
    
    ❯ gh api repos/ethereum-lists/chains/git/trees/4149a134bc500908ef28f43d32601b119473454b -q '{num_of_files: (.tree | length), truncated}'
    {
      "num_of_files": 2335,
      "truncated": false
    }
    ```

---

Note for Get a tree API endpoint:

> If truncated is true in the response then the number of items in the tree array exceeded our maximum limit. If you need to fetch more items, use the non-recursive method of fetching trees, and fetch one sub-tree at a time.
> 
> Note
> 
> The limit for the tree array is 100,000 entries with a maximum size of 7 MB when using the recursive parameter.

Use `recursive` query parameter for ~~laziness~~ convenience.

```
❯ gh api 'repos/ethereum-lists/chains/git/trees/fa19661238a17a36ee88e7cd9afebfa70becb5dc?recursive=1' -q '{num_of_files: (.tree | length), truncated}'
{
  "num_of_files": 4095,
  "truncated": false
}

❯ gh api 'repos/ethereum-lists/chains/git/trees/fa19661238a17a36ee88e7cd9afebfa70becb5dc?recursive=1' -q '[ .tree[] | select(.path | test("^_data/chains")) ][:3]'
[
  {
    "mode": "040000",
    "path": "_data/chains",
    "sha": "4149a134bc500908ef28f43d32601b119473454b",
    "type": "tree",
    "url": "https://api.github.com/repos/ethereum-lists/chains/git/trees/4149a134bc500908ef28f43d32601b119473454b"
  },
  {
    "mode": "100644",
    "path": "_data/chains/eip155-1.json",
    "sha": "647a94f5133c074bacc91453d09fe1184f073cc2",
    "size": 1664,
    "type": "blob",
    "url": "https://api.github.com/repos/ethereum-lists/chains/git/blobs/647a94f5133c074bacc91453d09fe1184f073cc2"
  },
  {
    "mode": "100644",
    "path": "_data/chains/eip155-10.json",
    "sha": "64c3c0388fcca1d7740cbf61ddb62d650bf278c3",
    "size": 1063,
    "type": "blob",
    "url": "https://api.github.com/repos/ethereum-lists/chains/git/blobs/64c3c0388fcca1d7740cbf61ddb62d650bf278c3"
  }
]
```
